File: frontend/src/api/speech.ts
export interface WordResult {
  text: string;
  start: number;
  end: number;
  duration: number;
}

export interface PauseResult {
  start: number;
  end: number;
  duration: number;
}

export interface SpeechAnalysisResult {
  transcript: string;
  recording_duration: number;
  speaking_duration: number;
  pause_duration: number;
  total_words: number;
  /** Legacy alias for speaking_words_per_minute. */
  words_per_minute: number;
  average_word_duration: number;
  words: WordResult[];
  /** WPM over the full recording duration. */
  overall_words_per_minute: number;
  /** WPM over detected speaking duration. */
  speaking_words_per_minute: number;
  pause_count: number;
  average_pause_duration: number;
  longest_pause_duration: number;
  silence_percentage: number;
  speech_percentage: number;
  filler_word_count: number;
  filler_word_rate: number;
  pace: string;
  fluency_score: number;
  pauses: PauseResult[];
}

const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8000";

export async function analyzeSpeech(
  audioBlob: Blob,
  filename: string,
): Promise<SpeechAnalysisResult> {
  const formData = new FormData();

  formData.append("file", audioBlob, filename);

  const response = await fetch(
    `${API_BASE_URL}/speech/analyze`,
    {
      method: "POST",
      body: formData,
    },
  );

  if (!response.ok) {
    let message = "Speech analysis failed.";

    try {
      const errorData: unknown = await response.json();

      if (
        typeof errorData === "object" &&
        errorData !== null &&
        "detail" in errorData &&
        typeof errorData.detail === "string"
      ) {
        message = errorData.detail;
      }
    } catch {
      // Keep the default error message.
    }

    throw new Error(message);
  }

  return (await response.json()) as SpeechAnalysisResult;
}

File: backend/app/schemas/speech.py
from pydantic import BaseModel, Field


class WordTiming(BaseModel):
    text: str
    start: float = Field(ge=0)
    end: float = Field(ge=0)
    duration: float = Field(ge=0)


class PauseTiming(BaseModel):
    start: float = Field(ge=0)
    end: float = Field(ge=0)
    duration: float = Field(ge=0)


class SpeechAnalysisResponse(BaseModel):
    transcript: str

    recording_duration: float = Field(ge=0)
    speaking_duration: float = Field(ge=0)
    pause_duration: float = Field(ge=0)

    total_words: int = Field(
        ge=0,
        description=(
            "Authoritative count of normalized transcribed words."
        ),
    )

    words_per_minute: float = Field(
        ge=0,
        description=(
            "Legacy alias for speaking_words_per_minute."
        ),
    )
    average_word_duration: float = Field(ge=0)

    words: list[WordTiming]

    overall_words_per_minute: float = Field(
        ge=0,
        description=(
            "Words per minute over the full recording duration."
        ),
    )
    speaking_words_per_minute: float = Field(
        ge=0,
        description=(
            "Words per minute over detected speaking duration."
        ),
    )
    pause_count: int = Field(ge=0)
    average_pause_duration: float = Field(ge=0)
    longest_pause_duration: float = Field(ge=0)
    silence_percentage: float = Field(ge=0)
    speech_percentage: float = Field(ge=0)
    filler_word_count: int = Field(ge=0)
    filler_word_rate: float = Field(ge=0)
    pace: str
    fluency_score: float = Field(ge=0, le=100)
    pauses: list[PauseTiming]
